File: Frameworks/conformance/src/load-spec.ts
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import type { OdsSpec } from './contract.ts'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

/**
 * Load a named spec from the shared `specs/` directory. Each call
 * returns a fresh object (no caching) so scenarios can't leak state
 * between runs.
 *
 * The Dart runner mirrors this helper — both sides read the exact
 * same JSON bytes, which is the whole point of the central directory.
 */
export function loadSpec(name: string): OdsSpec {
  const filepath = path.join(__dirname, '..', 'specs', `${name}.json`)
  const contents = fs.readFileSync(filepath, 'utf-8')
  return JSON.parse(contents) as OdsSpec
}

File: Frameworks/flutter-local/test/conformance/load_spec.dart
/// Loads a named spec from the shared `Frameworks/conformance/specs/`
/// directory. Mirrors `loadSpec` in
/// `Frameworks/conformance/src/load-spec.ts` — both sides read the same
/// JSON bytes, which is how we avoid spec divergence between drivers.
library;

import 'dart:convert';
import 'dart:io';

import 'contract.dart';

/// Path from the Flutter package root (CWD when `flutter test` runs) to the
/// shared specs directory.
const _specsDir = '../conformance/specs';

/// Load a named spec. Each call returns a fresh object (no caching) so
/// scenarios can't leak state between runs.
OdsSpec loadSpec(String name) {
  final file = File('$_specsDir/$name.json');
  if (!file.existsSync()) {
    throw StateError(
      'loadSpec: no spec file at ${file.path} '
      '(did you forget to create Frameworks/conformance/specs/$name.json?)',
    );
  }
  final content = file.readAsStringSync();
  return jsonDecode(content) as OdsSpec;
}
